layout/AddressesUsingIdsQuery.js

```diff
@@ -0,0 +1,138 @@
+'use strict';
+
+const _ = require('lodash');
+const Query = require('./Query');
+
+const baseQuery = {
+  query: {
+    function_score: {
+      query: {
+        bool: {
+          minimum_number_should_match: 1,
+          should: [],
+          filter: {
+            bool: {
+              minimum_number_should_match: 1
+            }
+          }
+        }
+      },
+      functions: []
+    }
+  }
+};
+
+function createShould(layer, ids) {
+  // create an object initialize with terms.'parent.locality_id' (or whatever)
+  return _.set({}, ['terms', `parent.${layer}_id`], ids);
+}
+
+function createAddressShould(vs) {
+  const should = {
+    bool: {
+      _name: 'fallback.address',
+      must: [
+        {
+          match_phrase: {
+            'address_parts.number': vs.var('input:housenumber').toString()
+          }
+        },
+        {
+          match_phrase: {
+            'address_parts.street': vs.var('input:street').toString()
+          }
+        }
+      ],
+      filter: {
+        term: {
+          layer: 'address'
+        }
+      }
+    }
+  };
+
+  if (vs.isset('boost:address')) {
+    should.bool.boost = vs.var('boost:address').toString();
+  }
+
+  return should;
+
+}
+
+function createStreetShould(vs) {
+  const should = {
+    bool: {
+      _name: 'fallback.street',
+      must: [
+        {
+          match_phrase: {
+            'address_parts.street': vs.var('input:street').toString()
```

Comment:
this `.toString()` convention is new to me.

there is a difference between using `vs.var('input:street')` and using `vs.var('input:street').toString()` in your views.

The former is a **lazy evaluation**, meaning that the value stored in the variable can change between instantiation of the query and execution of the query.

Using the latter syntax is an **eager evaluation**, meaning that the value stored in the variable is immediately converted to a scalar and subsequent changes to that variable will not be reflected in the generated query.

through the magic of js, when the object is marshalled using `JSON.stringify`, the `toString()` is called for you on all objects which support it, this magic is not super intuitive.

The advantage of lazy evaluation is that we can ensure a separation of concerns between the construction of a query layout, and the variables which are used to modify its behavior.

For complex queries it's also nice to be able to 'build' the query object once and then just simply substitute the variables for each execution of the query, this can actually save a fair amount of RAM.